src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -174,7 +165,20 @@ private Batch<Metric> Collect()
                         this.metrics[i].SnapShot();
                     }
 
-                    return new Batch<Metric>(this.metrics, indexSnapShot + 1);
+                    return Iterate(this.metrics, indexSnapShot + 1);
+
+                    // We cannot simply return the internal structure (array)
+                    // as the user is not expected to navigate it.
+                    // properly.
+                    static IEnumerable<Metric> Iterate(Metric[] metrics, long targetCount)
+                    {
+                        for (int i = 0; i < targetCount; i++)
+                        {
+                            // Check if the Metric has valid
+                            // entries and skip, if not.
+                            yield return metrics[i];
```

Comment:
I'm pretty sure yields always compile into an allocating structure.

Reply:
will come back to this.